browser/brave_shields/https_upgrade_browsertest.cc

```diff
@@ -0,0 +1,180 @@
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "base/test/scoped_feature_list.h"
+#include "brave/browser/brave_browser_process.h"
+#include "brave/components/brave_shields/browser/brave_shields_util.h"
+#include "brave/components/https_upgrade_exceptions/browser/https_upgrade_exceptions_service.h"
+#include "chrome/browser/browser_process.h"
+#include "chrome/browser/content_settings/host_content_settings_map_factory.h"
+#include "chrome/browser/interstitials/security_interstitial_page_test_utils.h"
+#include "chrome/browser/profiles/profile.h"
+#include "chrome/browser/ssl/https_only_mode_upgrade_interceptor.h"
+#include "chrome/common/pref_names.h"
+#include "chrome/test/base/chrome_test_utils.h"
+#include "components/prefs/pref_service.h"
+#include "content/public/test/browser_test.h"
+#include "content/public/test/browser_test_utils.h"
+#include "content/public/test/content_mock_cert_verifier.h"
+#include "content/public/test/test_navigation_observer.h"
+#include "net/cert/x509_certificate.h"
+#include "net/dns/mock_host_resolver.h"
+#include "third_party/blink/public/common/features.h"
+#include "url/gurl.h"
+
+#if BUILDFLAG(IS_ANDROID)
+#include "chrome/test/base/android/android_browser_test.h"
+#else
+#include "chrome/browser/ui/browser.h"
+#include "chrome/test/base/in_process_browser_test.h"
+#endif
+
+using blink::features::kHttpsByDefault;
+using brave_shields::ControlType;
+
+namespace {
+
+enum PageResult { HTTP, HTTPS, INTERSTITIAL };
+
+struct TestStruct {
+  bool init_secure;
+  const char* domain;
+  ControlType control_type;
+  PageResult expected_result;
+};
+
+const TestStruct test_combinations[] = {
+    {false, "insecure1.test", ControlType::ALLOW, PageResult::HTTP},
+    {false, "insecure2.test", ControlType::BLOCK_THIRD_PARTY, PageResult::HTTP},
+    {false, "insecure3.test", ControlType::BLOCK, PageResult::INTERSTITIAL},
+    {false, "upgradable1.test", ControlType::ALLOW, PageResult::HTTP},
+    {false, "upgradable2.test", ControlType::BLOCK_THIRD_PARTY,
+     PageResult::HTTPS},
+    {false, "upgradable3.test", ControlType::BLOCK, PageResult::HTTPS},
+    {true, "secure1.test", ControlType::ALLOW, PageResult::HTTPS},
+    {true, "secure2.test", ControlType::BLOCK_THIRD_PARTY, PageResult::HTTPS},
+    {true, "secure3.test", ControlType::BLOCK, PageResult::HTTPS}};
+
+#if BUILDFLAG(IS_ANDROID)
+base::FilePath GetChromeTestDataDir() {
+  return base::FilePath(FILE_PATH_LITERAL("net/data/url_request_unittest"));
+}
+#endif
+
+}  // namespace
+
+class HttpsUpgradeBrowserTest : public PlatformBrowserTest {
+ public:
+  HttpsUpgradeBrowserTest() = default;
+  ~HttpsUpgradeBrowserTest() override = default;
+
+  void SetUp() override {
+    feature_list_.InitAndEnableFeature(kHttpsByDefault);
+    PlatformBrowserTest::SetUp();
+  }
+
+  void SetUpOnMainThread() override {
+    g_brave_browser_process->https_upgrade_exceptions_service()
+        ->SetIsReadyForTesting();
```

Comment:
I would also like to see a test that uses data from this service.

Reply:
That is a good idea. @goodov, if it's all right with you, I'd like to write that test in a follow-up PR (soon). In the meantime @pes10k and I are hoping to merge this PR into Nightly as soon as possible. (Hoping this is OK as the feature is disabled by default and we will roll out gradually.)